### PR TITLE
Fix Django deprecation warnings (Fixes #2514)

### LINKED ISSUE #2514: Django 2.2 compatibility
The Travis CI builds already run against Django 2.2a1, so it might be good time to fix some of the warnings:

* [x] RemovedInDjango31Warning: Unit QuerySet won't use Meta.ordering in Django 3.1. Add .order_by('priority', 'position') to retain the current query. See [docs](https://docs.djangoproject.com/en/dev/releases/2.2/#model-meta-ordering-will-no-longer-affect-group-by-queries).
* [x] RemovedInDjango31Warning: `SimpleTestCase.allow_database_queries` is deprecated. Restrict the databases available during the execution of weblate.trans.tests.test_commands.BasicCommandTest with the `databases` attribute instead.

See https://travis-ci.org/WeblateOrg/weblate/jobs/482927553

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/69581844-django-2-2-compatibility?utm_campaign=plugin&utm_content=tracker%2F253393&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F253393&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by letitbeat**
Add the ordering property at model level to prevent the deprecation warning and order_by to required queries.

The allow_database_queries property is added if the Django version is lower than 3.x.x.

Signed-off-by: Jose Reyes <josereyesch@gmail.com>

### CODE REVIEW

File: weblate/checks/consistency.py [hunk 1]
```diff
@@ -89,7 +89,7 @@ def check_target_project(self, project):
             Count('target', distinct=True)
         ).filter(
             target__count__gt=1
-        )
+        ).order_by(*Unit.ordering)
```
Comment (nijel) [inline weblate/checks/consistency.py]: Ordering is not needed here.

File: weblate/checks/views.py [hunk 2]
```diff
@@ -161,7 +161,9 @@ def show_check_project(request, name, project):
             ).values(
                 'translation__component__slug',
                 'translation__component__project__slug'
-            ).annotate(count=Count('id'))
+            ).annotate(
+                count=Count('id')
+            ).order_by(*Unit.ordering)
```
Comment (nijel) [inline weblate/checks/views.py]: Ordering not needed here.

File: weblate/checks/views.py [hunk 3]
```diff
@@ -183,7 +185,9 @@ def show_check_project(request, name, project):
...
```
Comment (nijel) [inline weblate/checks/views.py]: Ordering not needed here.

File: weblate/memory/tasks.py [hunk 4]
```diff
@@ -53,7 +53,7 @@ def import_memory(project_id):
     units = Unit.objects.filter(
         translation__component__project_id=project_id,
         state__gte=STATE_TRANSLATED,
-    )
+    ).order_by(*Unit.ordering)
```
Comment (nijel) [inline weblate/memory/tasks.py]: Ordering not needed here.

File: weblate/trans/management/commands/__init__.py [hunk 5]
```diff
@@ -70,10 +70,10 @@ def add_arguments(self, parser):
     def get_units(self, **options):
         """Return list of units matching parameters."""
         if options['all']:
-            return Unit.objects.all()
+            return Unit.objects.all().order_by(*Unit.ordering)
         return Unit.objects.filter(
             translation__component__in=self.get_components(**options)
-        )
+        ).order_by(*Unit.ordering)
```
Comment (nijel) [inline weblate/trans/management/commands/__init__.py]: Ordering not needed here (and above).

File: weblate/trans/management/commands/rebuild_index.py [hunk 6]
```diff
@@ -94,7 +94,7 @@ def process_all(self, fulltext):
                         with writer.searcher() as searcher:
                             units = Unit.objects.filter(
                                 translation__language=language
-                            )
+                            ).order_by(*Unit.ordering)
```
Comment (nijel) [inline weblate/trans/management/commands/rebuild_index.py]: Ordering not needed here.

File: weblate/trans/models/component.py [hunk 7]
```diff
@@ -1807,7 +1808,7 @@ def run_target_checks(self):
         # Update has_failing_check flag
         allunits = Unit.objects.filter(
             translation__component__project=self.project
-        )
+        ).order_by(*Unit.ordering)
```
Comment (nijel) [inline weblate/trans/models/component.py]: Ordering not needed here.

File: weblate/trans/models/unit.py [hunk 8]
```diff
@@ -117,7 +117,7 @@ def filter_type(self, rqtype, project, language, ignored=False,
         elif rqtype == 'sourcecomments':
             coms = Comment.objects.filter(
                 language=None,
-            )
+            ).order_by(*Comment.ordering)
```
Comment (nijel) [inline weblate/trans/models/unit.py]: Ordering not needed here.

File: weblate/trans/models/unit.py [hunk 9]
```diff
@@ -621,7 +622,7 @@ def update_source_units(self, previous_source, user):
             id_hash=self.id_hash,
         ).exclude(
             id=self.id
-        )
+        ).order_by(*Unit.ordering)
```
Comment (nijel) [inline weblate/trans/models/unit.py]: Ordering not needed here. Used only in update query later.

File: weblate/trans/tasks.py [hunk 10]
```diff
@@ -237,7 +239,7 @@ def cleanup_suggestions():
                 target=suggestion.target
             ).exclude(
                 id=suggestion.id
-            )
+            ).order_by(*Suggestion.ordering)
```
Comment (nijel) [inline weblate/trans/tasks.py]: Ordering not needed here (used only for delete later).

File: weblate/trans/tests/test_commands.py [hunk 11]
```diff
@@ -345,7 +347,10 @@ def test_import_mercurial_mixed(self):
 
 
 class BasicCommandTest(SimpleTestCase):
-    allow_database_queries = True
+    databases = '__all__'
+
+    if not re.match(r"3.[0-9].[0-9]", django.get_version()):
+        allow_database_queries = True
```
Comment (nijel) [inline weblate/trans/tests/test_commands.py]: Does this have to be behind if? The additional attribute should not matter and this just brings complex statement into several places...

File: weblate/trans/views/basic.py [hunk 12]
```diff
@@ -357,7 +361,10 @@ def new_language(request, project, component):
 
         if form.is_valid():
             langs = form.cleaned_data['lang']
-            for language in Language.objects.filter(code__in=langs):
+            languages = Language.objects.filter(
+                code__in=langs
+            ).order_by(*Language.ordering)
```
Comment (nijel) [inline weblate/trans/views/basic.py]: Ordering not needed here.

File: weblate/trans/tests/test_commands.py (hunk 11 above)
Comment (letitbeat) [reply weblate/trans/tests/test_commands.py]: Agree with you regarding the complexity but without the "if" we cannot get rid of the Warnings.

File: weblate/trans/models/translation.py [hunk 13]
```diff
@@ -101,10 +101,10 @@ class Translation(models.Model, URLMixin, LoggerMixin):
     objects = TranslationManager.from_queryset(TranslationQuerySet)()
 
     is_lockable = False
-    _reverse_url_name = 'translation'
+    _reverse_url_name = 'translation'
```
Comment (comradekingu) [inline weblate/trans/models/translation.py]:
```suggestion
    _reverse_url_name = 'translation'
```

File: weblate/trans/views/changes.py [hunk 14]
```diff
@@ -204,7 +204,8 @@ def get_queryset(self):
 
         if self.user is not None:
             result = result.filter(user=self.user)
-
+        
+        result = result.order_by(*Change.ordering)
```
Comment (comradekingu) [inline weblate/trans/views/changes.py]:
```suggestion
        result = result.order_by(*Change.ordering)
```

**Comment by nijel** [comment]
I'm sorry to say that, but I'm not going to merge this as pretty much all queries need to be investigated and considered for ordering in this case. You've missed lot of cases where ordering is needed (admin, last changes to name some) and in some cases the ordering is added when needed (I've pointed out some in the review, but apparently I've missed some). Also in many situations the ordering can be simplified compared to current setup.

Overall this is something what needs deeper knowledge of current codebase. But anyway thanks a lot for your effort.

PS: I'm currently going through this and slowly fixing the issue (see 988aadc00b342415f28bc86d21fb13930e8197d4 or bb84857eb05fc66c41436508a84ec4971eb2daee).

**Comment by letitbeat** [comment]
Sure thing, I understand, most changes were added because of new warnings triggered by Django in pagination and init methods from the rest framework for the lack of default ordering in querysets, let me know if do you want me to open a new PR for the `allow_database_queries` change.

**Comment by nijel** [comment]
Thanks, that one has been also addressed differently in 308fedba6dfae6b87402908695049d428ebeaaa3.

**Pull request closed without merging**